cellar_rebalancer/src/allocation.rs

```diff
@@ -0,0 +1,157 @@
+use crate::{cellars::uniswapv3::UniswapV3CellarTickInfo, prelude::*, somm_send};
+use abscissa_core::Application;
+use deep_space::{Coin, Contact};
+use ethers::prelude::U256;
+use proto::query_client::QueryClient as AllocationQueryClient;
+use somm_proto::somm as proto;
+use std::time::Duration;
+use tokio::time::{sleep, timeout};
+use tonic::transport::Channel;
+
+pub struct Connections {
+    pub grpc: Option<AllocationQueryClient<Channel>>,
+    pub contact: Option<Contact>,
+}
+
+pub async fn allocation_precommit(
+    cosmos_key: &deep_space::private_key::PrivateKey,
+    allocation: &proto::Allocation,
+    pair_id: String,
+) -> proto::AllocationPrecommit {
+    let config = APP.config();
+    let delegate_cosmos_address = cosmos_key
+        .to_address(&config.cosmos.prefix)
+        .unwrap()
+        .to_string();
+    let hasher = somm_send::data_hash(allocation, delegate_cosmos_address)
+        .await
+        .unwrap();
+    proto::AllocationPrecommit {
+        hash: hasher.hash,
+        cellar_id: pair_id,
+    }
+}
+
+pub async fn cellar_query_client() -> Connections {
```

Comment:
This provides both the `AllocationQueryClient` and the `Contact` for sending Cosmos transactions, so a more generic name might be appropriate.